sw/device/lib/base/random_order.c

```diff
@@ -5,15 +5,61 @@
 #include "sw/device/lib/base/random_order.h"
 
 #include "sw/device/lib/base/bitfield.h"
-
-// TODO: The current implementation is just a skeleton, and currently just
-// traverses from 0 to `min_len * 2`.
+#include "sw/device/lib/base/hardened.h"
 
 void random_order_init(random_order_t *ctx, size_t min_len) {
-  ctx->state = 0;
-  ctx->max = min_len * 2;
+  // Position relative to start index is always initialized to zero.
+  ctx->pos = 0;
+
+  // If the sequence length is zero or one, there's not much randomization we
+  // can do, but we can at least return one value past the range.
+  if (launder32(min_len) < 2) {
+    ctx->state = 0;
+    ctx->step = 1;
+    ctx->max = min_len + 1;
+    return;
+  }
+  HARDENED_CHECK_LE(2, min_len);
+
+  // Sample the start index from the range [0, n-1].
+  ctx->state = random_order_random_word() % min_len;
+  // Sample the step size from the range [1, n-1].
+  ctx->step = (random_order_random_word() % (min_len - 1)) + 1;
+
+  if (min_len - (min_len % ctx->step) > UINT32_MAX - ctx->step) {
+    // Safe fallback for cases where the max calculation would overflow.
+    ctx->max = UINT32_MAX;
+    ctx->step = 1;
+  } else {
+    // The maximum value is the smallest multiple of `step` that is strictly
+    // greater than `min_len`.
+    ctx->max = min_len - (min_len % ctx->step) + ctx->step;
+  }
+  HARDENED_CHECK_LE(min_len, ctx->max);
 }
 
 size_t random_order_len(const random_order_t *ctx) { return ctx->max; }
 
-size_t random_order_advance(random_order_t *ctx) { return ctx->state++; }
+size_t random_order_advance(random_order_t *ctx) {
```

Comment:
Something that may be interesting from a fault detection perspective is to have a counter increment on each `random_order_advance` call. Then at the end we can check if the number of increments matches our expectation. 

Is there another way we can use to ensure that there were no indexes skipped during any given iteration?

Reply:
Hmm, this is a good idea. I think I can tweak the way the internal tracking works to do this, and add another call that's `random_order_assert_done` or something and does a hardened check against the length.

Reply:
Pushed a new commit now to implement this.